src/main/java/graphql/normalized/NormalizedQueryTreeFactory.java

```diff
@@ -113,34 +144,301 @@ private void fixUpParentReference(NormalizedField rootNormalizedField) {
+        }
+        ImmutableList<String> objectTypeNames = map(objectTypes, GraphQLObjectType::getName);
+        NormalizedField normalizedField = NormalizedField.newQueryExecutionField()
+                .alias(field.getAlias())
+                .resolvedArguments(argumentValues)
+                .normalizedArguments(normalizedArgumentValues)
+                .astArguments(field.getArguments())
+                .objectTypeNames(objectTypeNames)
+                .fieldName(fieldName)
+                .level(level)
+                .parent(parent)
+                .build();
+
+        result.put(resultKey, normalizedField);
+        fieldsByNormalizedField.put(normalizedField, field);
+    }
+
+    private NormalizedField findMatchingNF(GraphQLSchema schema, Collection<NormalizedField> normalizedFields, GraphQLFieldDefinition fieldDefinition, List<Argument> arguments) {
+        for (NormalizedField nf : normalizedFields) {
+            GraphQLFieldDefinition nfFieldDefinition = nf.getOneFieldDefinition(schema);
+            // same field name
+            if (!nfFieldDefinition.getName().equals(fieldDefinition.getName())) {
+                continue;
+            }
+            // same type
+            if (!simplePrint(nfFieldDefinition.getType()).equals(simplePrint(fieldDefinition.getType()))) {
+                continue;
+            }
+            // same arguments
+            if (!sameArguments(nf.getAstArguments(), arguments)) {
+                continue;
+            }
+            return nf;
+        }
+        return null;
+    }
+
+    // copied from graphql.validation.rules.OverlappingFieldsCanBeMerged
+    private boolean sameArguments(List<Argument> arguments1, List<Argument> arguments2) {
+        if (arguments1.size() != arguments2.size()) {
+            return false;
+        }
+        for (Argument argument : arguments1) {
+            Argument matchedArgument = findArgumentByName(argument.getName(), arguments2);
+            if (matchedArgument == null) {
+                return false;
+            }
+            if (!sameValue(argument.getValue(), matchedArgument.getValue())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Argument findArgumentByName(String name, List<Argument> arguments) {
+        for (Argument argument : arguments) {
+            if (argument.getName().equals(name)) {
+                return argument;
+            }
+        }
+        return null;
+    }
+
+    private boolean sameValue(Value value1, Value value2) {
+        if (value1 == null && value2 == null) {
+            return true;
+        }
+        if (value1 == null) {
+            return false;
+        }
+        if (value2 == null) {
+            return false;
+        }
+        return new AstComparator().isEqual(value1, value2);
```

Comment:
could this code go in AstComparator() -

Also does AstComparator need to be an object allocation?  Why not a static method

Reply:
great point ... moved